### ISSUE TITLE
"reference to packed field is unaligned" in println, but not when the same arg is used elsewhere

### ISSUE THREAD

**Issue by osa1** [rust-lang/rust#87946]
Tried with current master (ccffcafd55e).

This builds without warnings:

```rust
#[repr(packed)]
pub struct ObjInd {
    _header: u32,
    field: u32,
}

pub unsafe fn test(obj: *const ObjInd) {
    let field = (*obj).field;
    println!("{}", field);
}
```

But this generates a "ref to packed field" warning:

```rust
#[repr(packed)]
pub struct ObjInd {
    _header: u32,
    field: u32,
}

pub unsafe fn test(obj: *const ObjInd) {
    println!("{}", (*obj).field);
}
```

Output:

```
   Compiling packed_field_ptr v0.1.0 (/home/omer/rust/packed_field_ptr)
warning: reference to packed field is unaligned
 --> src/lib.rs:8:20
  |
8 |     println!("{}", (*obj).field);
  |                    ^^^^^^^^^^^^
  |
  = note: `#[warn(unaligned_references)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #82523 <https://github.com/rust-lang/rust/issues/82523>
  = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)

warning: `packed_field_ptr` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
```

Same with similar macros like `writeln`.

My guess is `println` is taking reference of its arguments, to avoid moving. No idea how to fix these warnings or if it's even possible without breaking backwards compat by making `println` move the args.

**Comment by RalfJung** [comment]
Yes, `assert_eq!` also adds an implicit `&`.

**Comment by fralalonde** [comment]
Quick workaround: here's an `align_copy!` macro that copies the field value to a temp var and returns the now-aligned value copy. If there's a nicer way to do this, I'd like to know.

    macro_rules! align_copy {
        ($value:expr) => {{ let _copy = $value; _copy }};
    }
    
    // example usage
    println!("{}", align_copy!(packed_struct.unaligned_field))

**Comment by RalfJung** [comment]
`println!("{}", { packed_struct.unaligned_field })` should work without the macro.

I am renaming this issue to track printing a better error message for code like this
```rust
#[repr(packed)]
pub struct ObjInd {
    _header: u32,
    field: u32,
}

pub unsafe fn test(obj: *const ObjInd) {
    println!("{}", (*obj).field);
}
```
It is expected that this code gets rejected, but the error could explain that there is an implicit `&` added by the format macro, and that one way to resolve the issue is create a copy of the field data by writing `println!("{}", { (*obj).field });`

**Comment by fralalonde** [comment]
Doesn't the `(*obj).field` form imply that the whole struct will be copied first? This seems wasteful if a single field needs to be accessed out of a large struct, more so if it's multiple fields, each requiring a full copy. Unless the full copy can be reliably optimized away by the compiler to keep only the field value? In which case, the macro approach would not be required.

**Comment by RalfJung** [comment]
No, that code means that only the field is copied. The place-to-value conversion happens after the field projection.
